hnix-store-core/src/System/Nix/Internal/Nar/Parser.hs

```diff
@@ -0,0 +1,446 @@
+expectRawString expected = do
+  actual <- consume (BS.length expected)
+  when (actual /= expected) $
+    Fail.fail $ "Expected " ++ err expected ++ ", got " ++ err actual
+  where
+    err bs =
+      if   BS.length bs > 10
+      then show (BS.take 10 bs) ++ "..."
+      else show bs
+
+
+-- | Consume a NAR string, and dispatch to a parser depending on which string
+--   matched
+matchStr
+  :: (IO.MonadIO m, Fail.MonadFail m)
+  => [(T.Text, NarParser m a)]
+     -- ^ List of expected possible strings and the parsers they should run
+  -> NarParser m a
+matchStr parsers = do
+  str <- parseStr
+  case List.lookup str parsers of
+    Just p  -> p
+    Nothing -> Fail.fail $ "Expected one of " ++ show (fst <$> parsers) ++ " found " ++ show str
+
+
+-- | Wrap any parser in NAR formatted parentheses
+--   (a parenthesis is a NAR string, so it needs length encoding and padding)
+parens :: (IO.MonadIO m, Fail.MonadFail m) => NarParser m a -> NarParser m a
+parens act = do
+  expectStr "("
+  r <- act
+  expectStr ")"
+  return r
+
+
+-- | Sort links in the symlink stack according to their connectivity
+--   (Targets must be created before the links that target them)
+createLinks :: IO.MonadIO m => NarParser m ()
+createLinks = do
+  createLink  <- Reader.asks Nar.narCreateLink
+  allLinks    <- State.gets links
+  sortedLinks <- IO.liftIO $ sortLinksIO allLinks
+  flip mapM_ sortedLinks $ \li -> do
+    pwd <- IO.liftIO $ Directory.getCurrentDirectory
+    IO.liftIO $ Directory.setCurrentDirectory (linkPWD li)
+    Trans.lift $ createLink (linkTarget li) (linkFile li)
+    IO.liftIO $ Directory.setCurrentDirectory pwd
+
+      where
+
+        -- Convert every target and link file to a filepath relative
+        -- to NAR root, then @Graph.topSort@ it, and map from the
+        -- relative filepaths back to the original @LinkInfo@.
+        -- Relative paths are needed for sorting, but @LinkInfo@s
+        -- are needed for creating the link files
+        sortLinksIO :: [LinkInfo] -> IO [LinkInfo]
+        sortLinksIO ls = do
+          linkLocations <- fmap Map.fromList $
+            forM ls $ \li->
+                        (,li) <$> Directory.canonicalizePath (linkFile li)
+          canonicalLinks <- forM ls $ \l -> do
+            targetAbsPath <- Directory.canonicalizePath
+                             (linkPWD l </> linkTarget l)
+            fileAbsPath   <- Directory.canonicalizePath
+                             (linkFile l)
+            return (fileAbsPath, targetAbsPath)
+          let linkGraph = Graph.edges canonicalLinks
+          case Graph.topSort linkGraph of
+            Nothing          -> error "Symlinks form a loop"
+            Just sortedNodes ->
+              let
+                sortedLinks = flip Map.lookup linkLocations <$> sortedNodes
+              in
+                return $ catMaybes sortedLinks
+
+
+------------------------------------------------------------------------------
```

Comment:
Some of the `State manipulation` guys don't need to live at the top level, move under `where` clauses.

Reply:
Nah